### PR TITLE
Made benchmark application flexible to run separate benchmarks. Chang…

### PR DESCRIPTION
**Pull request by LepilkinaElena**
…e structure to reuse steps for other benchmark applications.

### CODE REVIEW

File: performance/build.gradle [hunk 1]
```diff
@@ -17,140 +17,46 @@ buildscript {
...
-konanRun.finalizedBy konanJsonReport
-
-private void dumpReport(String name, ByteArrayOutputStream output) {
-    new File("${name}").withOutputStream {
-        it.write(output.toByteArray())
-    }
+// Produce and send slack report.
+task slackReport(type: RegressionsReporter) {
+    // Create folder for report (root Kotlin project settings make create report in separate folder).
+    def reportDirectory = outputReport.substring(0, outputReport.lastIndexOf("/"))
```
Comment (olonho) [inline performance/build.gradle]: If it is JVM, `File(outputReport).parentFile` may be better.

File: performance/build.gradle [hunk 2]
```diff
@@ -17,140 +17,46 @@ buildscript {
...
+    mkdir reportDirectory
+    mkdir "../targetsResults"
```
Comment (olonho) [inline performance/build.gradle]: Same here, better use platform provided mechanisms of getting parent directories and such.

File: performance/buildSrc/src/main/kotlin/MPPTools.kt [hunk 3]
```diff
@@ -110,6 +110,17 @@ fun createJsonReport(projectProperties: Map<String, Any>): String {
...
+    val reportsToMerge = reports.map {
+        val json = it.inputStream().bufferedReader().use { it.readText() }
+        val reportElement = JsonTreeParser.parse(json)
+        BenchmarksReport.create(reportElement)
+
+    }
+    return reportsToMerge.reduce { result, it -> result + it }.toJson()
+}
+
+
```
Comment (olonho) [inline performance/buildSrc/src/main/kotlin/MPPTools.kt]: Extra line.

File: performance/buildSrc/src/main/kotlin/RunKotlinNativeTask.kt [hunk 4]
```diff
@@ -18,6 +20,10 @@ open class RunKotlinNativeTask @Inject constructor(
     var buildType = "RELEASE"
     var workingDir: Any = project.projectDir
     var outputFileName: String? = null
+    @Input
+    @Option(option="subkernel", description="subkernel")
```
Comment (olonho) [inline performance/buildSrc/src/main/kotlin/RunKotlinNativeTask.kt]: option ="subkernel"

Comment (olonho) [reply performance/buildSrc/src/main/kotlin/RunKotlinNativeTask.kt]: What is subkernel?

File: performance/gradle/benchmark.gradle [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    }
+}
+
+MPPTools.addTimeListener(project)
+
+MPPTools.createRunTask(project, 'konanRun', kotlin.targets.native) {
+    workingDir = project.provider {
+        kotlin.targets.native.compilations.main.getBinary('EXECUTABLE', buildType).parentFile
+    }
+    args("-w", "$nativeWarmup", "-r", "$attempts", "-o", "${buildDir.absolutePath}/${nativeBenchResults}", "-p", "${project.ext.applicationName}::")
```
Comment (olonho) [inline performance/gradle/benchmark.gradle]: What is `::` here?

File: performance/gradle/benchmark.gradle [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+task konanJsonReport {
+    doLast {
+        def nativeExecutable = MPPTools.getKotlinNativeExecutable(kotlin.targets.native, "RELEASE")
+        def nativeCompileTime = MPPTools.getNativeCompileTime(project.ext.applicationName)
+        String benchContents = new File("${buildDir.absolutePath}/${nativeBenchResults}").text
+        def properties = getCommonProperties() + ['type'           : 'native',
+                                                  'compilerVersion': "${konanVersion}".toString(),
+                                                  'flags'          : kotlin.targets.native.compilations.main.extraOpts.collect{ "\"$it\"" },
```
Comment (olonho) [inline performance/gradle/benchmark.gradle]: Strange indent.

File: performance/ring/src/main/kotlin/org/jetbrains/ring/StringBenchmark.kt [hunk 7]
```diff
@@ -22,7 +22,7 @@ open class StringBenchmark {
         get() = _data!!
     var csv: String = ""
 
-    fun setup() {
+    init {
```
Comment (olonho) [inline performance/ring/src/main/kotlin/org/jetbrains/ring/StringBenchmark.kt]: Why not make it more Kotlin'sh while you're here?

File: performance/shared/src/main/kotlin-jvm/org/jetbrains/benchmarksLauncher/Utils.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.io.File
+
+actual fun writeToFile(fileName: String, text: String) {
+    File(fileName).printWriter().use { out ->
+        out.println(text)
+    }
+}
+
+// Wrapper for assert funtion in stdlib
```
Comment (olonho) [inline performance/shared/src/main/kotlin-jvm/org/jetbrains/benchmarksLauncher/Utils.kt]: Comments end with the dot.

File: performance/buildSrc/src/main/kotlin/RunKotlinNativeTask.kt (hunk 4 above)
Comment (LepilkinaElena) [reply performance/buildSrc/src/main/kotlin/RunKotlinNativeTask.kt]: This is set of separate benchmarks from benchmark application. I don't know what name would be better for this.

File: performance/gradle/benchmark.gradle (hunk 5 above)
Comment (LepilkinaElena) [reply performance/gradle/benchmark.gradle]: It's just end of prefix to have something like "Ring::ClassArray.allocateArray" in report.

File: performance/build.gradle [hunk 9]
```diff
@@ -17,140 +17,47 @@ buildscript {
...
-}
-
-jvmRun.finalizedBy jvmJsonReport
-konanRun.finalizedBy konanJsonReport
-
-private void dumpReport(String name, ByteArrayOutputStream output) {
-    new File("${name}").withOutputStream {
-        it.write(output.toByteArray())
-    }
+// Produce and send slack report.
```
Comment (olonho) [inline performance/build.gradle]: slack -> Slack.

File: performance/build.gradle [hunk 10]
```diff
@@ -168,35 +75,44 @@ private def uploadBenchmarkResultToBintray(String fileName) {
     }
 }
 
-task bench(type:Exec) {
-    dependsOn jvmRun
-    dependsOn konanRun
+def mergeReports(String fileName) {
```
Comment (olonho) [inline performance/build.gradle]: Maybe better rewrite in Kotlin?

File: performance/cinterop/src/main/kotlin-native/org/jetbrains/cinteropBenchmarks/cinteropBenchmark.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.jetbrains.cinteropBenchmarks
+import kotlinx.cinterop.*
+
+const val benachmarkSize = 10000
+
+actual fun macrosBenchmark() {
+    memScoped {
+        val ints = new_list_int()
+        for (i in 1..benachmarkSize) {
```
Comment (olonho) [inline performance/cinterop/src/main/kotlin-native/org/jetbrains/cinteropBenchmarks/cinteropBenchmark.kt]: `benachmarkSize` -> `benchmarkSize`

**Pull request closed without merging**